code/modules/ninja/energy_katana.dm

```diff
@@ -1,30 +1,87 @@
-/obj/item/weapon/katana/energy
+/obj/item/weapon/dash
+	name = "energy katana"
+	var/max_charges = 3
+	var/current_charges = 3
+	var/charge_rate = 30 //In deciseconds
+	var/dash_toggled = FALSE
+
+/obj/item/weapon/dash/proc/charge()
+	current_charges = Clamp(current_charges + 1, 0, max_charges)
+	if(istype(loc, /mob/living))
+		to_chat(loc, "<span class='notice'>[src] now has [current_charges]/[max_charges] charges.</span>")
+
+/obj/item/weapon/dash/attack_self(mob/user)
+	dash_toggled = !dash_toggled
+	if(dash_toggled)
+		to_chat(loc, "<span class='notice'>[src] is now in dash mode.</span>")
+	else
+		to_chat(loc, "<span class='notice'>You disable the dash function on [src].</span>")
+
+/obj/item/weapon/dash/afterattack(atom/target, mob/user, proximity_flag, click_parameters)
+	if(dash_toggled)
+		dash(user, target)
+		return
+
+/obj/item/weapon/dash/proc/dash(mob/user, atom/target)
+	if(!current_charges)
+		return
+
+	if(isliving(target))
+		return
+
+	if(target.density)
+		return
+
+	var/turf/T = get_turf(target)
+	if(target in view(user.client.view, get_turf(user)))
+		var/obj/spot1 = new /obj/effect/temp_visual/dir_setting/ninja/phase/out(get_turf(user), user.dir)
+		user.forceMove(T)
+		playsound(T, 'sound/magic/blink.ogg', 25, 1)
+		playsound(T, "sparks", 50, 1)
+		var/obj/spot2 = new /obj/effect/temp_visual/dir_setting/ninja/phase(get_turf(user), user.dir)
+		var/chain = spot1.Beam(spot2,"blur",time=15)
+		current_charges--
+		addtimer(CALLBACK(src, .proc/charge), charge_rate)
+		sleep(2)
```

Comment:
also set ```chain```'s time to ```2```, instead of ```15```, instead of qdeling it after 2, that way lines 45-48 can be removed.